[feat] add favicons to result urls

vojkovic opened the pull request:
## What does this PR do?

<!-- MANDATORY -->
Adds favicons near the urls in results.

![image](https://github.com/user-attachments/assets/9304bd1c-9eb1-4199-95c2-b13422572775)
![image](https://github.com/user-attachments/assets/759b9677-09b5-45f8-aeec-1484efa60ea0)
![image](https://github.com/user-attachments/assets/bf582b4c-b0b4-48ad-b9c5-203171a3f0ad)
![image](https://github.com/user-attachments/assets/e9669b0b-841b-490b-af1f-34a3bb06b642)
(Spacing is broken on RTL mode due to https://github.com/searxng/searxng/issues/3724)

This works very similar to the implementation of the autocomplete backends.

You can also turn off favicons completely in results by setting it to none. I think some optimisations to my css and templates can be made for rtl mode.

After my tests, I believe that caching on the client-side is decent enough, so I've not implemented any cache to SearXNG. Let me know your feedback.

You can see the changes live already on https://priv.au, although this also has my theme patches.

## Related issues

Closes: https://github.com/searxng/searxng/issues/1014 (after 2 years!)

- https://github.com/searxng/searxng/issues/1014#issuecomment-1079044776

vojkovic committed:
[feat] add favicons to result urls

return42 commented on searx/webapp.py:
```diff
@@ -866,6 +888,42 @@ def autocompleter():
+        # using redirect to save network bandwidth (user will have this location cached).
+        return redirect(url_for('static', filename='themes/' + theme + '/img/empty_favicon.svg'))
+
+    # will always return a PNG image
```
OK, here is my first question: I don't think we should assume that all favicons are PNGs .. when I try https://f1.allesedv.com/32/google.com for example, I see that it is an `image/x-icon` mime-type ... or do you see it differently than I do?

vojkovic replied:
Yup - it's an x-icon on allesedv. It still works so it slipped by me. Maybe we can implement a way of just copying the media type of the backend.

return42 commented on searx/static/themes/simple/img/empty_favicon.svg:
Next question :-) .. from where does this symbol (SVG) comes from / who is the creator, do we have the rights to use it or are there any copyrights? 

Does this symbol have a special meaning?

vojkovic replied:
Not really - it's just one of the first globe icons I saw. It's MIT licenced, from https://www.svgrepo.com/svg/507325/globe-alt

return42 commented on searx/static/themes/simple/src/less/search.less:
```diff
@@ -378,3 +378,12 @@ html.no-js #clear_search.hide_if_nojs {
+  height: 1.8rem;
+  width: 1.8rem;
+  border-radius: 20%;
+  background-color: #ddd;
```
Did you have any special reason to set the background color? Some favicons have a transparency; I personally would prefer to leave the background of the `<img>` transparent so that the background of the theme shines through the favicon. We already have a border that provides a visual demarcation.  In your opinion, would there be anything against leaving the background of the `<img>` transparent and deleting the background-color?

vojkovic replied:
Yeah, there's a good reason to have a white background colour.

When first testing the feature out, I had your idea as well. It's just some icons aren't made for no background, see:
<img width="718" alt="nobackground" src="https://github.com/user-attachments/assets/a0356f9b-7396-4597-ac99-f8c0a1dabe49">
vs:
<img width="702" alt="withbackground" src="https://github.com/user-attachments/assets/aba03a89-3137-45f9-a3f8-7eafbd0e817d">

return42 replied:
Yeah, makes sense .. thanks for clarify :+1:

Bnyro commented on searx/webapp.py:
```diff
@@ -297,6 +298,24 @@ def morty_proxify(url: str):
+
+    h = new_hmac(settings['server']['secret_key'], url.encode())
+
+    return '{0}?{1}'.format(url_for('favicon_proxy'), urlencode(dict(q=url.encode(), h=h)))
```
```suggestion
    return f"{url_for('favicon_proxy')}?{urlencode(dict(q=url.encode(), h=h)))}"
```

return42 replied:
@Bnyro thanks for your review, but please wait a little bit with this PR / I'm in review with this PR .. I modified a lot .. I will need some more time to finish my rework .. Then you can make a review.

Bnyro commented on searx/webapp.py:
```diff
@@ -866,6 +888,42 @@ def autocompleter():
+        return '', 400
+
+    # parse query
+    raw_text_query = RawTextQuery(url, [])
```
I don't think using a `RawTextQuery` makes sense here since we're dealing with an fqdn, not a user query input?

Bnyro commented on searx/templates/simple/macros.html:
```diff
@@ -21,9 +21,29 @@
 {% macro result_header(result, favicons, image_proxify) -%}
 <article class="result {% if result['template'] %}result-{{ result.template|replace('.html', '') }}{% else %}result-default{% endif %} {% if result['category'] %}category-{{ result['category'] }}{% endif %}{% for e in result.engines %} {{ e }}{% endfor %}">
   {{- result_open_link(result.url, "url_wrapper") -}}
+  {% if not rtl %}
```
I don't really like the duplicated code for rtl/ltr. Can't we just do that via CSS by changing the flex direction?

return42 commented on searx/static/themes/simple/img/empty_favicon.svg:
```diff
@@ -0,0 +1,5 @@
+<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24">
+  <path fill="#fff" d="M0 0h24v24H0z"/>
```
Here we can delete the white background in the SVG icon (transparent). As mentioned in https://github.com/searxng/searxng/pull/3727#discussion_r1718513261, the `<img>` background in the theme is already designed for transparent icons.

vojkovic replied:
This was an intentional design decision by me, although it's up to your preference, but I prefer this icon as a bright white (#fff) rather than the dull white of the regular background (#ddd).

return42 committed:
[mod] sqlitedb: access to SQLite databases a little more convenient.

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] Revision of the favicon solution

All favicons implementations have been documented and moved to the Python
package:

    searx.favicons

There is a configuration (based on Pydantic) for the favicons and all its
components:

    searx.favicons.config

A solution for caching favicons has been implemented:

    searx.favicon.cache

If the favicon is already in the cache, the returned URL is a data URL [1]
(something like `data:image/png;base64,...`).  By generating a data url from
the FaviconCache, additional HTTP roundtripps via the favicon_proxy are saved:

    favicons.proxy.favicon_url

The favicon proxy service now sets a HTTP header "Cache-Control: max-age=...":

    favicons.proxy.favicon_proxy

The resolvers now also provide the mime type (data, mime):

    searx.favicon.resolvers

[1] https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] UI: add favicon's border- and background color to the definitons

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

vojkovic commented on searx/favicons/resolvers.py:
```diff
@@ -0,0 +1,81 @@
+    # URL https://www.google.com/s2/favicons?sz=32&domain={domain}" will be
+    # redirected (HTTP 301 Moved Permanently) to t1.gstatic.com/faviconV2:
+    response = network.get(
+        f"https://t1.gstatic.com/faviconV2?client=SOCIAL&type=FAVICON&url=https://{domain}&size=32",
```
Looks like this only works with some websites.

Doesn't work:
https://t1.gstatic.com/faviconV2?client=SOCIAL&type=FAVICON&url=https://reddit.com&size=32

For it to load needs the `fallback_opts` param:
https://t0.gstatic.com/faviconV2?client=SOCIAL&type=FAVICON&fallback_opts=TYPE,SIZE,URL&url=http://reddit.com&size=32


But some (actually most) sites work fine as:
https://t1.gstatic.com/faviconV2?client=SOCIAL&type=FAVICON&url=https://priv.au&size=32

and from my brief testing, they also do with the `fallback_opts` param, so maybe this should be included:

https://t1.gstatic.com/faviconV2?client=SOCIAL&type=FAVICON&fallback_opts=TYPE,SIZE,URL&url=https://priv.au&size=32

return42 replied:
> For it to load needs the `fallback_opts` param:

Yeah this is also what the redirect https://www.google.com/s2/favicons?sz=32&domain=reddit.com does .. thought I can shorten the URL a little bit .. but I can't :-)

Fixed and squashed into the WIP commit on top of the branch [diff](https://github.com/searxng/searxng/compare/e63e83b1d1688201f338b01a0187c5a3c162ef7f..f3acc085b709a9aba45ddb801a85466d56d1fbe4).

vojkovic commented on searx/static/themes/simple/src/less/definitions.less:
```diff
@@ -227,6 +230,9 @@
   --color-doc-code-background: #4d5a6f;
+  // Favicons Colors
+  --color-favicon-background-color: #bbb;
+  --color-favicon-border-color: #333;
```
Is there a reason as to why we're changing the background colours in light and dark mode?

return42 replied:
I found the background of transparent favicons (and the border) a bit too bright in dark mode ... so I thought it would be good to include this background and the border into the `definitions.less`.  I then made the two colors less bright in the dark theme.  If you think the latter is not good we can use the same colors in light and dark theme .. but we should not remove theses both two colors from the `definitions.less`.

vojkovic replied:
I'll leave this at your discretion - but I think the same colors should be used on both themes. There's not enough contrast between the favicon and background on the dark theme in my opinion.

return42 replied:
No problem .. now we have same colors in dark & light theme / [diff](https://github.com/searxng/searxng/compare/bbfd9249ef47e5afc7ebd7aca5d1300606959f69..388cdcad2c94ad0d05fb6757125ffc15d9c7272f)

return42 commented on searx/favicons/__init__.py:
```diff
@@ -0,0 +1,37 @@
+    if not cfg_file.exists():
+        if is_active():
+            logger.error(f"missing favicon config: {cfg_file}")
+        cfg_file = config.DEFAULT_CFG_TOML
```
Config file is now read from `/etc/searxng/favicons.toml`

return42 commented on searx/templates/simple/macros.html:
```diff
@@ -21,9 +21,19 @@
   {{- result_open_link(result.url, "url_wrapper") -}}
+  {% if not rtl %}
+    {%- if favicon_resolver != "" %}
+    <div class="favicon"><img loading="lazy" src="{{ favicon_url(result.parsed_url.netloc) }}"></div>
```
favicons are now loaded lazy (will save some bandwidth)

vojkovic replied:
Looks good - I was never able to get it working, but do you think we can avoid the rtl formatting in the html?

return42 replied:
I have not yet found a solution (TBH I have not yet dealt it) .. @Bnyro suggests CSS in https://github.com/searxng/searxng/pull/3727#pullrequestreview-2240912152 .. I'm not sure how.

return42 commented on searx/sqlitedb.py:
```diff
@@ -0,0 +1,322 @@
+        #
+        # if self._DB is None:
+        #     self._DB = self.connect()  # pylint: disable=attribute-defined-outside-init
+        # return self._DB
```
I had a lot of trouble with the optimization of DB connections in Python 3.12 ..  https://docs.python.org/3/library/sqlite3.html#sqlite3.threadsafety says

> 3. Serialized: In serialized mode, SQLite can be safely used by multiple threads with no restriction.
> .. Threads may share the module, connections and cursors

But I have seen lots of issues when sharing the DB cursor over multiple threads.

to avoid this problem, one DB connection per thread is now set up for “read-only accesses” / a new connection had to be created for SQL updates, as the SQL updates in the different threads would otherwise block each other (as long as a transaction is running in one thread, the other thread may not make any updates, these competing updates could only be synchronized with separate DB connections)

return42 reviewed (commented):
I did some more tests today and fixed some bugs and made some minor improvements ... I will create the documentation in the next few days.

Here are a few notes on my work from today ...

vojkovic commented on searx/static/themes/simple/src/less/style-ltr.less:
```diff
@@ -82,4 +82,8 @@
+.favicon {
+  margin: 0 8px 0 0;
+}
+
```
When I wrote this, I was sure there must be a better way - but nothing I did worked, so I just kept it to discuss later. Can we use flex to maybe fix this?

return42 committed:
[doc] documentation of the favicons infrastructure

Run ``make docs.live`` and visit http://0.0.0.0:8000/admin/searx.favicons.html

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

vojkovic commented on docs/admin/searx.favicons.rst:
```diff
@@ -0,0 +1,154 @@
+:py:obj:`cache.db_url <.FaviconCacheConfig.db_url>`:
+  The path to the (SQLite_) database file.  The default path is in the `/tmp`_
+  folder, which is deleted on every reboot and is therefore unsuitable for a
+  productive environment.  The FHS_ provides the folder  for the
```
'productive environment' -> 'production environment'

vojkovic commented on docs/admin/searx.favicons.rst:
```diff
@@ -0,0 +1,249 @@
+
+.. sidebar:: warning
+
+   Don't activate the favacions before reading the documentation.
```
favacions -> favicons

vojkovic commented on docs/admin/searx.favicons.rst:
```diff
@@ -0,0 +1,249 @@
+parts:
+
+- :py:obj:`Favicons-Proxy <.favicons.proxy>` (aka *proxy*)
+- :py:obj:`Favicons-Resolvers <.favicons.resolvers>` (aka *reolver*)
```
reolver -> resolver

vojkovic commented on docs/admin/settings/settings_search.rst:
```diff
@@ -41,6 +42,11 @@
+``favicon_resolver``:
+  To activate favicons in SearXNG's result list select a default
+  favicon-resolver, leave blank to turn off the feature.  Don't activate the
+  favacions before reading the :ref:`Favicons documentation <favicons>`.
```
favacions -> favicons

vojkovic commented on docs/admin/searx.favicons.rst:
```diff
@@ -0,0 +1,249 @@
+     $ sudo chmod g+w /var/cache/searxng/
+
+:py:obj:`cache.LIMIT_TOTAL_BYTES <.FaviconCacheConfig.LIMIT_TOTAL_BYTES>`:
+  Maximum of bytes (default) stored in the cache of all blobs. The limit is only
```
Maximum of bytes (default)?

Did you forget to include the figure?

vojkovic commented on docs/admin/searx.favicons.rst:
```diff
@@ -0,0 +1,249 @@
+including a *cache*.  The cache must be parameterized according to your own
+requirements and maintained regularly.
+
+To activate favicons in SearXNG's result list set a default favicons-resolver in
```
Add comma:

To activate favicons in SearXNG's result list set a default favicons-resolver in

To activate favicons in SearXNG's result list, set a default favicons-resolver in

vojkovic commented on docs/admin/searx.favicons.rst:
```diff
@@ -0,0 +1,249 @@
+   search:
+     favicon_resolver: "duckduckgo"
+
+By default and without any extensions SearXNG serves these resolvers:
```
Add comma:

By default and without any extensions SearXNG serves these resolvers:

By default and without any extensions, SearXNG serves these resolvers:

Bnyro commented on docs/admin/searx.favicons.rst:
```diff
@@ -0,0 +1,249 @@
+
+Activating the favicons in SearXNG is very easy, but this **generates a
+significantly higher load** in the communication between client and server and
+additional resources are used on the server.
```
```suggestion
significantly higher load** in the communication between client and server and
increases resource usage on the server.
```
nit: two `and`s, grammar

return42 replied:
my intention was to say: *"higher load in the client/server communication and increases resource usage on the server."* --> [fixed](https://github.com/searxng/searxng/compare/a76b37fe75f50c8281fbec24499e74211ba2b67e..7c7482fc107484e883803df65176647d58aff989)

Bnyro commented on docs/admin/searx.favicons.rst:
```diff
@@ -0,0 +1,249 @@
+including a *cache*.  The cache must be parameterized according to your own
+requirements and maintained regularly.
+
+To activate favicons in SearXNG's result list, set a default favicons-resolver
```
```suggestion
To activate favicons in SearXNG's result list, set a default ``favicon_resolver``
```
named favicon_resolver without s below

Bnyro commented on docs/admin/searx.favicons.rst:
```diff
@@ -0,0 +1,249 @@
+With the above setting favicons are displayed, the user has the option to
+deactivate this feature in his settings.  If the user is to have the option of
+selecting from several *resolvers*, a further setting is required / *but this will
+be discussed later, first we have to setup the favicons cache*.
```
```suggestion
With the above setting, favicons are displayed by default. Users have the option to
deactivate this feature in their settings..
```
Forgot to remove? / TODO?

return42 replied:
Nope .. "but this will be discussed later" means later in this article .. but I [fixed](https://github.com/searxng/searxng/compare/a76b37fe75f50c8281fbec24499e74211ba2b67e..7c7482fc107484e883803df65176647d58aff989) this part to be more clear.

Bnyro commented on searx/favicons/config.py:
```diff
@@ -0,0 +1,62 @@
+CONFIG_SCHEMA: int = 1
+"""Version of the configuration schema."""
+
+TOML_CACHE: dict[str, "FaviconConfig"] = {}
```
```suggestion
TOML_CACHE_CFG: dict[str, "FaviconConfig"] = {}
```

return42 replied:
[fixed](https://github.com/searxng/searxng/compare/7c7482fc107484e883803df65176647d58aff989..2021ebc593f7fea2c2c893bb0361f8ff860d3f13)

Bnyro commented on searx/favicons/config.py:
```diff
@@ -0,0 +1,62 @@
+TOML_CACHE: dict[str, "FaviconConfig"] = {}
+"""Cache config objects by TOML's filename."""
+
+DEFAULT_CFG_TOML = pathlib.Path(__file__).parent / "favicons.toml"
```
```suggestion
DEFAULT_CFG_TOML_PATH = pathlib.Path(__file__).parent / "favicons.toml"
```

Bnyro reviewed (commented):
Everything looks fine to me generally, just left some small comments above.

I didn't deeply review the database part though as I'm not that familiar with SQL in Python.

return42 committed:
[mod] favicons: simplify RTL in template & CSS

Comes from a sughgestion in:

- https://github.com/searxng/searxng/pull/3727#issuecomment-2388998803

Suggested-by: Bnyro <bnyro@tutanota.com>
Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[build] /static
